docs/architecture/adr-019-protobuf-state-encoding.md

```diff
@@ -110,162 +111,155 @@ type Codec interface {
+usage of `Any`, it just introduces a security and introspection layer for 
+golang usage.
 
-  amino *codec.Codec
-}
-
-func NewAppCodec(amino *codec.Codec) *Codec {
-  return &Codec{Marshaler: codec.NewHybridCodec(amino), amino: amino}
-}
+### Using `Any` to encode state
 
-func (c *Codec) MarshalAccount(accI authexported.Account) ([]byte, error) {
-  acc := &Account{}
-  if err := acc.SetAccount(accI); err != nil {
-    return nil, err
-  }
+The SDK will provide support methods `MarshalAny` and `UnmarshalAny` to allow
+easy encoding of state to `Any` in `Codec` implementations. Ex:
 
-  return c.Marshaler.MarshalBinaryBare(acc)
+```go
+func (c *Codec) MarshalEvidence(evidenceI eviexported.Evidence) ([]byte, error) {
+	return codec.MarshalAny(evidenceI)
 }
 
-func (c *Codec) UnmarshalAccount(bz []byte) (authexported.Account, error) {
-  acc := &Account{}
-  if err := c.Marshaler.UnmarshalBinaryBare(bz, acc); err != nil {
-    return nil, err
-  }
-
-  return acc.GetAccount(), nil
+func (c *Codec) UnmarshalEvidence(bz []byte) (eviexported.Evidence, error) {
+	var evi eviexported.Evidence
+	err := codec.UnmarshalAny(c.interfaceContext, &evi, bz)
+	if err != nil {
+		return nil, err
+	}
+	return evi, nil
 }
 ```
 
-Since the `Codec` implements `auth.Codec` (and all other required interfaces), it is passed to _all_
-the modules and satisfies all the interfaces. Now each module needing to work with interfaces will know
-about all the required types. Note, the use of `interface_type` allows us to avoid a significant
-amount of code boilerplate when implementing the `Codec`.
-
-A similar concept is to be applied for messages that contain interfaces fields. The module will
-define a "base" concrete message type that the application-level codec will extend via `oneof` that
-fulfills the required message interface.
-
-Example:
-
-The `MsgSubmitEvidence` defined by the `x/evidence` module contains a field `Evidence` which is an
-interface.
-
-```go
-type MsgSubmitEvidence struct {
-  Evidence  exported.Evidence
-  Submitter sdk.AccAddress
-}
-```
+### Using `Any` in `sdk.Msg`s
 
-Instead, we will implement a "base" message type and an interface which the concrete message type
-must implement.
+A similar concept is to be applied for messages that contain interfaces fields.
+For example, we can define `MsgSubmitEvidence` as follows where `Evidence` is
+an interface:
 
 ```protobuf
 // x/evidence/types/types.proto
 
-message MsgSubmitEvidenceBase {
+message MsgSubmitEvidence {
   bytes submitter = 1
     [
       (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"
     ];
+  google.protobuf.Any evidence = 2;
 }
 ```
 
-```go
-// x/evidence/exported/evidence.go
+Note that in order to unpack the evidence from `Any` we do need a reference to
+`InterfaceContext`. This is inconvenient if we want to reference the evidence
+in methods like `ValidateBasic`. We can work around this limitation by
```

Comment:
How will this work in places like `ValidateBasic` or other methods that take no arguments and are stateless?

Reply:
The description that follows explains this (basically the `UnpackInterfaces` phase). Is it not clear enough or does the solution need to be introduced earlier?

Reply:
I see. I had to re-read it a few times. Essentially, we're relying on `UnpackInterfaces` to be called _directly_ after unmarshaling? If so, what happens when we don't. Methods like `ValidateBasic` will just panic? Is there a call we can make in functions like `ValidateBasic` to make sure `UnpackInterfaces` was called?

Reply:
Not without relying on a global registry. `ValidateBasic` could return an `error` instead of panicking of course

Also note that `UnpackInterfaces` isn't needed if you just called `NewMsgSubmitEvidence` in a CLI method. The real value will be cached immediately when the msg is constructed. `UnpackInterfaces` will only be needed when unmarshaling from binary or JSON.

Reply:
I've tried to explain things better here: https://github.com/cosmos/cosmos-sdk/pull/6081/commits/6db53b2d46fe19238d79417de1745295f9b88587

Let me know if this is sufficient or needs improvement.

Reply:
> ValidateBasic could return an error instead of panicking of course.

Yes, I'm strongly in favor of this. These methods and functions need not panic.

Yes, that commit helps clarify things 👍